pkg/protocol/block_dispatcher.go

```diff
@@ -0,0 +1,345 @@
+	dispatchWorkers *workerpool.WorkerPool
+
+	// warpSyncWorkers is the worker pool that is used to process the WarpSync requests and responses.
+	warpSyncWorkers *workerpool.WorkerPool
+
+	// unsolidCommitmentBlocks is a buffer that stores blocks that have an unsolid slot commitment.
+	unsolidCommitmentBlocks *buffer.UnsolidCommitmentBuffer[*types.Tuple[*model.Block, network.PeerID]]
+
+	// pendingWarpSyncRequests is the set of pending requests that are waiting to be processed.
+	pendingWarpSyncRequests *eventticker.EventTicker[iotago.SlotIndex, iotago.CommitmentID]
+
+	// processedWarpSyncRequests is the set of processed requests.
+	processedWarpSyncRequests ds.Set[iotago.CommitmentID]
+
+	// shutdownEvent is a reactive event that is triggered when the BlockDispatcher instance is stopped.
+	shutdownEvent reactive.Event
+
+	// optWarpSyncWindowSize is the optional warp sync window size.
+	optWarpSyncWindowSize iotago.SlotIndex
+}
+
+// NewBlockDispatcher creates a new BlockDispatcher instance.
+func NewBlockDispatcher(protocol *Protocol, opts ...options.Option[BlockDispatcher]) *BlockDispatcher {
+	return options.Apply(&BlockDispatcher{
+		protocol:                  protocol,
+		dispatchWorkers:           protocol.Workers.CreatePool("BlockDispatcher.Dispatch"),
+		warpSyncWorkers:           protocol.Workers.CreatePool("BlockDispatcher.WarpSync", 1),
+		unsolidCommitmentBlocks:   buffer.NewUnsolidCommitmentBuffer[*types.Tuple[*model.Block, network.PeerID]](20, 100),
+		pendingWarpSyncRequests:   eventticker.New[iotago.SlotIndex, iotago.CommitmentID](eventticker.RetryInterval[iotago.SlotIndex, iotago.CommitmentID](WarpSyncRetryInterval)),
+		processedWarpSyncRequests: ds.NewSet[iotago.CommitmentID](),
+		shutdownEvent:             reactive.NewEvent(),
+	}, opts, func(b *BlockDispatcher) {
+		protocol.HookConstructed(b.initEngineMonitoring)
+		protocol.HookInitialized(b.initNetworkConnection)
+		protocol.HookStopped(b.shutdown)
+	})
+}
+
+// Dispatch dispatches the given block to the correct engine instance.
+func (b *BlockDispatcher) Dispatch(block *model.Block, src network.PeerID) error {
+	slotCommitment := b.protocol.ChainManager.LoadCommitmentOrRequestMissing(block.ProtocolBlock().SlotCommitmentID)
+	if !slotCommitment.SolidEvent().WasTriggered() {
+		if !b.unsolidCommitmentBlocks.Add(slotCommitment.ID(), types.NewTuple(block, src)) {
+			return ierrors.Errorf("failed to add block %s to unsolid commitment buffer", block.ID())
+		}
+
+		return ierrors.Errorf("failed to dispatch block %s: slot commitment %s is not solid", block.ID(), slotCommitment.ID())
+	}
+
+	matchingEngineFound := false
+	for _, engine := range []*engine.Engine{b.protocol.MainEngineInstance(), b.protocol.CandidateEngineInstance()} {
+		if engine != nil && (engine.ChainID() == slotCommitment.Chain().ForkingPoint.ID() || engine.BlockRequester.HasTicker(block.ID())) {
+			if !b.inWarpSyncRange(engine, block) {
+				engine.ProcessBlockFromPeer(block, src)
+			}
```

Comment:
Why don't we forward this block to the engine? This basically means if the block is too far ahead we will simply discard it? If it is to prevent future blocks to bloat the memory, then it would be very nice to just put a comment here that explains exactly that.

Reply:
I renamed the method to "inSyncWindow" and added a corresponding comment in the method description.